hdf5::write fixes

bglaessle opened the pull request:
Hi.

I am using your wrapper for hdf5 output (mostly non
lattice data) from chroma. Thanks for providing it.

Up to this date these are a few fixes which were critical
to get my code running. And they should make
debugging for any user a lot easier. Read the commit
messages for more details.

I am also happy to adjust my patches, if your unhappy
with them.

Regards, Benjamin

ps. I have also other additions (I enabled chunking in most
cases and added manual checksums) in case ur interested.

Commit:
added globalness value and size checks (cirit

if this is not done I have often hard to trace
HDF5 faults. In theory the interface user could
do them, but in practive it is very often not
transparent (e.g. in a Chroma environment) what
is global and what is not and way easier to do
it here.

Similar checks might also be relevant in write routines,
I have only implemented them in the most basic cases!

Commit:
check max attribute size of 64kB

there is a way to get around it, but I
wasnt sure about it and its simply easier
to check.
Otherwise attributes that are to large
lead to more hdf5 errors

Commit:
made H5Awrite calls collective again (reverts 0a3a7eb2)

I had problems that attributes existed but had zero values,
apparently H5Awrite needs to collective

Commit:
fixed fatal out of memory bug, reduced memory footprint of multi[12]d writes

If multiple mpi process on the same node allocate a buffer simultaneously
(on Linux) and this exceeds the memory available, Linux will happily
allow overallocation (all processes get "valid" pointers) and the Job
will crash with untraceable segfaults.

FIX: The whole thing is unnecessary from the start since we only need
a buffer on one node (=0).

The bug is also present in specialized writes but far less likely to occur
because of the smaller (static) sizes. This is NOT fixed right now!
It is also present in LatticeObject writes, but there almost impossible
to catch. (also not fixed)

Commit:
reverted to single c++98 constructor, fixed missing return

some intel compiler didnt appreciate the new c++
standard too much.

azrael417 commented:
Hi Benjamin,

thanks for you patches. I will review them asap. I also applied big changes, I.e. prefetch the cb->Lexi maps once and then perform the local reordering using an OpenMP parallel loop. This should enhance BGQ performance significantly. But I think that change is orthogonal to what you applied. However, I am curious why some changes were critical. I tested a lot of cases and it worked fine for me. What was the problem here? I was the only user so far, therefore I am happy that someone else is trying it out.

Best
Thorsten 

> Am 11.01.2015 um 04:48 schrieb bglaessle notifications@github.com:
> 
> Hi.
> 
> I am using your wrapper for hdf5 output (mostly non
> lattice data) from chroma. Thanks for providing it.
> 
> Up to this date these are a few fixes which were critical
> to get my code running. And they should make
> debugging for any user a lot easier. Read the commit
> messages for more details.
> 
> I am also happy to adjust my patches, if your unhappy
> with them.
> 
> Regards, Benjamin
> 
> ps. I have also other additions (I enabled chunking it most
> cases and added manual checksums) in case ur interested.
> 
> You can merge this Pull Request by running
> 
>   git pull https://github.com/bglaessle/qdpxx azrael-fixes
> Or view, comment on, or merge it at:
> 
>   https://github.com/azrael417/qdpxx/pull/1
> 
> Commit Summary
> 
> added globalness value and size checks (cirit
> check max attribute size of 64kB
> made H5Awrite calls collective again (reverts 0a3a7eb2)
> fixed fatal out of memory bug, reduced memory footprint of multi[12]d writes
> reverted to single c++98 constructor, fixed missing return
> File Changes
> 
> M include/qdp_hdf5.h (146)
> M lib/qdp_hdf5.cc (87)
> Patch Links:
> 
> https://github.com/azrael417/qdpxx/pull/1.patch
> https://github.com/azrael417/qdpxx/pull/1.diff
> —
> Reply to this email directly or view it on GitHub.

bglaessle commented:
Hi.

Thanks for the fast answer.

I didn't touch any of the LatticeObject output routines, so I probably havent
interfered with what you were doing.

I had a lot of problems with write calls that were not "properly collective":
Sometimes variables did not have the same value/size on all nodes, which
resulted in crashs when closing the file. (Mostly the size is critical,
for INDEPENDENT dataset writes this should be enough, but for attributes
because of the issue below, I am also checking values).

Also apparently H5Awrite needs to be collective (I did not revert the other
changes of that commit), because in one of my test the data of these attributes
wasnt properly written out,  when not - they were zero/emtpy instead.

Another problem was that I was almost running out of RAM (for other reasons)
and I was writing out large arrays (~250MB). When all jobs on 1 node
simultaenously allocated the write buffer, Linux granted all of them (theoretically),
but this will overallocate, which results in SEGFAULTS with no sensible error message as
soon as I touch some part of this memory.
This was some almost unavoidable race condition as far as I understood it.

The other thing were rather minor: compiler dependencies and checking for max
attribute size.

The commit messages also contain some information.

Sometimes my error handling is certainly extrem and rather verbose,
but this can be adjusted.

Do you have some kind of testing environment, unit tests, etc
which you run by default? I must admit my testing only covered my use
cases and was in no way "elaborate" or general.

Regards, Benjamin

ps. I might allow this to be linked against scalar hdf5 librarys. Are you in
    principal interested in merging such changes back in?

Am 11.01.2015 um 15:16 schrieb Thorsten Kurth:

> Hi Benjamin,
> 
> thanks for you patches. I will review them asap. I also applied big changes, I.e. prefetch the
> cb->Lexi maps once and then perform the local reordering using an OpenMP parallel loop. This
> should enhance BGQ performance significantly. But I think that change is orthogonal to what you
> applied. However, I am curious why some changes were critical. I tested a lot of cases and it
> worked fine for me. What was the problem here? I was the only user so far, therefore I am happy
> that someone else is trying it out.
> 
> Best
> Thorsten
> 
> > Am 11.01.2015 um 04:48 schrieb bglaessle notifications@github.com:
> > 
> > Hi.
> > 
> > I am using your wrapper for hdf5 output (mostly non
> > lattice data) from chroma. Thanks for providing it.
> > 
> > Up to this date these are a few fixes which were critical
> > to get my code running. And they should make
> > debugging for any user a lot easier. Read the commit
> > messages for more details.
> > 
> > I am also happy to adjust my patches, if your unhappy
> > with them.
> > 
> > Regards, Benjamin
> > 
> > ps. I have also other additions (I enabled chunking it most
> > cases and added manual checksums) in case ur interested.
> > 
> > You can merge this Pull Request by running
> > 
> > git pull https://github.com/bglaessle/qdpxx azrael-fixes
> > Or view, comment on, or merge it at:
> > 
> > https://github.com/azrael417/qdpxx/pull/1
> > 
> > Commit Summary
> > 
> > added globalness value and size checks (cirit
> > check max attribute size of 64kB
> > made H5Awrite calls collective again (reverts 0a3a7eb2)
> > fixed fatal out of memory bug, reduced memory footprint of multi[12]d writes
> > reverted to single c++98 constructor, fixed missing return
> > File Changes
> > 
> > M include/qdp_hdf5.h (146)
> > M lib/qdp_hdf5.cc (87)
> > Patch Links:
> > 
> > https://github.com/azrael417/qdpxx/pull/1.patch
> > https://github.com/azrael417/qdpxx/pull/1.diff
> > —
> > Reply to this email directly or view it on GitHub.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/azrael417/qdpxx/pull/1#issuecomment-69495863.
